tcmur_cmd_handler.c

```diff
@@ -456,6 +456,150 @@ static int handle_read(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+	state->offset += state->length;
+	if (state->done_blocks < dev->num_lbas) {
+		if ((dev->num_lbas - state->done_blocks) * dev->block_size < state->length) {
+			state->length = (dev->num_lbas - state->done_blocks) * dev->block_size;
+			free_iovec(writecmd);
+			if (alloc_iovec(writecmd, state->length)) {
+				ret = tcmu_set_sense_data(sense, HARDWARE_ERROR,
+							  ASC_INTERNAL_TARGET_FAILURE,
+							  NULL);
+				goto free_cmd;
+			}
+		}
+
+		tcmu_err("next format cmd, done_blocks:%lu num_lbas:%lu block_size:%lu\n",
+			 state->done_blocks, dev->num_lbas, dev->block_size);
+		rc = async_handle_cmd(dev, writecmd, format_unit_work_fn);
+		if (rc != TCMU_ASYNC_HANDLED) {
+			ret = tcmu_set_sense_data(sense, MEDIUM_ERROR,
+						  ASC_WRITE_ERROR,
+						  NULL);
+			goto free_iovec;
+		}
+		state->done_blocks += state->length / dev->block_size;
+		if (state->done_blocks < dev->num_lbas)
+			dev->format_progress = (0x10000 * state->done_blocks) /
+						dev->num_lbas;
+	}
+
+	if (state->done_blocks == dev->num_lbas) {
+		tcmu_err("last format cmd, done_blocks:%lu num_lbas:%lu block_size:%lu\n",
+			 state->done_blocks, dev->num_lbas, dev->block_size);
+		goto free_iovec;
+	}
+	return;
+
+free_iovec:
+	free_iovec(writecmd);
+free_cmd:
+	free(writecmd);
+	free(state);
+	pthread_mutex_lock(&dev->format_lock);
+	dev->flags &= ~FORMATTING;
+	pthread_mutex_unlock(&dev->format_lock);
+	aio_command_finish(dev, origcmd, ret);
+}
+
+static int handle_format_unit(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	struct tcmulib_cmd *writecmd;
+	struct format_unit_state *state;
+	size_t length = 1024 * 1024;
+	uint8_t *sense = cmd->sense_buf;
+	int ret;
+
+	pthread_mutex_lock(&dev->format_lock);
+	if (dev->flags & FORMATTING) {
+		pthread_mutex_unlock(&dev->format_lock);
+		return tcmu_set_sense_data(sense, NOT_READY,
+					  ASC_NOT_READY_FORMAT_IN_PROGRESS,
+					  &dev->format_progress);
+	}
+	dev->format_progress = 0;
+	dev->flags |= FORMATTING;
+	pthread_mutex_unlock(&dev->format_lock);
+
+	writecmd = calloc(1, sizeof(*writecmd));
+	if (!writecmd)
+		return SAM_STAT_TASK_SET_FULL;
+	writecmd->done = handle_format_unit_cbk;
+	writecmd->cmdstate = cmd;
+
+	state = calloc(1, sizeof(*state));
+	if (!state)
+		goto free_iov;
+
+	cmd->cmdstate = state;
+	state->done_blocks = 0;
+	state->length = length;
+
+	if ((dev->num_lbas - state->done_blocks) * dev->block_size < length)
+		state->length = (dev->num_lbas - state->done_blocks) * dev->block_size;
+
+	if (alloc_iovec(writecmd, state->length)) {
+		free(state);
+		goto free_cmd;
+	}
+
+	tcmu_err("start emulate format, done_blocks:%lu num_lbas:%lu block_size:%lu\n",
+		 state->done_blocks, dev->num_lbas, dev->block_size);
+
+	/* copy incase handler changes it */
+	state->write_buf = writecmd->iovec->iov_base;
+
+	ret = async_handle_cmd(dev, writecmd, format_unit_work_fn);
+	if (ret != TCMU_ASYNC_HANDLED)
+		goto free_state;
+
+	state->done_blocks += state->length / dev->block_size;
```

Comment:
I had this all done at the beginning of the call back before, but this still doesn't explain why its not doing any writes when callback function is calling async_handle_cmd.